File: P7_Ensambles/practica7_ensambles_alumnos.py
"""
PRÁCTICA 7: La Inteligencia Colectiva (Modelos de Ensambles)
Asignatura: Introducción al Aprendizaje Automático (IAA) 26/27

Código base para el alumnado.

Objetivo del script
-------------------
Este archivo proporciona una estructura guiada para que el alumnado implemente
la práctica paso a paso. No es una solución completa: contiene secciones TODO
que deben completarse y analizarse en el informe final.

Dataset utilizado
-----------------
Breast Cancer Wisconsin (incluido en scikit-learn).

Modelos a trabajar
------------------
1. Árbol de decisión simple
2. Random Forest
3. Gradient Boosting (o AdaBoost, si se quiere probar como alternativa)

Recomendación
-------------
Lee cada comentario antes de programar. No se trata solo de obtener métricas,
sino de interpretar los resultados y justificar lo que observas.
"""

from __future__ import annotations

import time
from typing import Dict, List, Tuple

import matplotlib.pyplot as plt
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier


RANDOM_STATE = 42
TEST_SIZE = 0.25


# -----------------------------------------------------------------------------
# 1. CARGA Y PREPARACIÓN DE LOS DATOS
# -----------------------------------------------------------------------------
def load_dataset() -> Tuple[pd.DataFrame, pd.Series]:
    """
    Carga el dataset Breast Cancer Wisconsin desde scikit-learn.

    Returns
    -------
    X : pd.DataFrame
        Variables predictoras.
    y : pd.Series
        Etiqueta de clase.
    """
    data = load_breast_cancer()
    X = pd.DataFrame(data.data, columns=data.feature_names)
    y = pd.Series(data.target, name="target")
    return X, y


def split_dataset(
    X: pd.DataFrame,
    y: pd.Series,
    test_size: float = TEST_SIZE,
    random_state: int = RANDOM_STATE,
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:
    """
    Divide el dataset en entrenamiento y prueba.

    TODO para el alumnado:
    - Revisa por qué usamos stratify=y.
    - Comprueba cuántas muestras hay en entrenamiento y en test.
    """
    return train_test_split(
        X,
        y,
        test_size=test_size,
        random_state=random_state,
        stratify=y,
    )


# -----------------------------------------------------------------------------
# 2. TAREA 1: ÁRBOL SIMPLE (BASELINE)
# -----------------------------------------------------------------------------
def train_decision_tree(
    X_train: pd.DataFrame,
    y_train: pd.Series,
    random_state: int = RANDOM_STATE,
) -> DecisionTreeClassifier:
    """
    Entrena un árbol de decisión sin limitar la profundidad.

    TODO para el alumnado:
    - Comprueba qué ocurre cuando no fijamos max_depth.
    - Relaciona el resultado con el concepto de sobreajuste.
    """
    model = DecisionTreeClassifier(random_state=random_state)
    model.fit(X_train, y_train)
    return model


def evaluate_model(
    model,
    X_train: pd.DataFrame,
    y_train: pd.Series,
    X_test: pd.DataFrame,
    y_test: pd.Series,
) -> Dict[str, float]:
    """
    Evalúa un modelo en entrenamiento y en prueba.

    Returns
    -------
    results : dict
        Diccionario con accuracy en train y test.

    TODO para el alumnado:
    - Añade otras métricas si lo consideras útil.
    - Interpreta la diferencia entre train_accuracy y test_accuracy.
    """
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    train_accuracy = accuracy_score(y_train, y_train_pred)
    test_accuracy = accuracy_score(y_test, y_test_pred)

    return {
        "train_accuracy": train_accuracy,
        "test_accuracy": test_accuracy,
    }


# -----------------------------------------------------------------------------
# 3. TAREA 2: RANDOM FOREST
# -----------------------------------------------------------------------------
def random_forest_experiment(
    X_train: pd.DataFrame,
    y_train: pd.Series,
    X_test: pd.DataFrame,
    y_test: pd.Series,
    n_trees_list: List[int],
    random_state: int = RANDOM_STATE,
) -> pd.DataFrame:
    """
    Ejecuta varios experimentos cambiando el número de árboles.

    Parameters
    ----------
    n_trees_list : list of int
        Lista con los valores de n_estimators que se desean probar.

    Returns
    -------
    results_df : pd.DataFrame
        Resultados del experimento.

    TODO para el alumnado:
    - Observa si llega un punto de saturación.
    - Compara estos resultados con el árbol simple.
    """
    rows = []

    for n_trees in n_trees_list:
        start = time.perf_counter()

        model = RandomForestClassifier(
            n_estimators=n_trees,
            random_state=random_state,
        )
        model.fit(X_train, y_train)

        elapsed = time.perf_counter() - start
        y_test_pred = model.predict(X_test)
        test_accuracy = accuracy_score(y_test, y_test_pred)

        rows.append(
            {
                "n_estimators": n_trees,
                "test_accuracy": test_accuracy,
                "train_time_seconds": elapsed,
            }
        )

    return pd.DataFrame(rows)


def plot_random_forest_results(results_df: pd.DataFrame) -> None:
    """
    Representa la evolución del accuracy en test frente al número de árboles.

    TODO para el alumnado:
    - Cambia el gráfico para mostrar error en lugar de accuracy si lo prefieres.
    - Interpreta la forma de la curva en el informe.
    """
    plt.figure(figsize=(8, 5))
    plt.plot(
        results_df["n_estimators"],
        results_df["test_accuracy"],
        marker="o",
        linestyle="--",
    )
    plt.title("Random Forest: Accuracy en test vs número de árboles")
    plt.xlabel("n_estimators")
    plt.ylabel("Accuracy en test")
    plt.grid(True)
    plt.tight_layout()
    # Save the plot to the specified path for the report
    import os
    save_path = "/home/jesus/IAA/docLaTex/sections/practica7/img/rf_accuracy.png"
    os.makedirs(os.path.dirname(save_path), exist_ok=True)
    plt.savefig(save_path)
    print(f"Plot saved successfully to {save_path}")


# -----------------------------------------------------------------------------
# 4. TAREA 3: BOOSTING
# -----------------------------------------------------------------------------
def train_gradient_boosting(
    X_train: pd.DataFrame,
    y_train: pd.Series,
    random_state: int = RANDOM_STATE,
) -> Tuple[GradientBoostingClassifier, float]:
    """
    Entrena un modelo Gradient Boosting y mide el tiempo de entrenamiento.

    TODO para el alumnado:
    - Si quieres, sustituye este modelo por AdaBoost y compara ambos.
    - Interpreta la diferencia entre entrenamiento secuencial (boosting)
      y entrenamiento por ensamble de árboles más independientes (bagging).
    """
    start = time.perf_counter()

    model = GradientBoostingClassifier(random_state=random_state)
    model.fit(X_train, y_train)

    elapsed = time.perf_counter() - start
    return model, elapsed


# -----------------------------------------------------------------------------
# 5. IMPORTANCIA DE VARIABLES
# -----------------------------------------------------------------------------
def show_top_features(model, feature_names: List[str], top_k: int = 3) -> pd.DataFrame:
    """
    Muestra las variables más importantes según el modelo.

    Parameters
    ----------
    model : estimator
        Modelo ya entrenado con atributo feature_importances_.
    feature_names : list of str
        Nombres de las variables.
    top_k : int, default=3
        Número de variables a mostrar.

    Returns
    -------
    top_features : pd.DataFrame
        Tabla con las variables más importantes.

    TODO para el alumnado:
    - Comenta si estas variables tienen sentido en el problema.
    - Compara la importancia obtenida en Random Forest y en Boosting.
    """
    importances = pd.DataFrame(
        {
            "feature": feature_names,
            "importance": model.feature_importances_,
        }
    )
    importances = importances.sort_values("importance", ascending=False)
    return importances.head(top_k)


# -----------------------------------------------------------------------------
# 6. UTILIDADES DE IMPRESIÓN
# -----------------------------------------------------------------------------
def print_confusion_and_report(model, X_test: pd.DataFrame, y_test: pd.Series) -> None:
    """
    Imprime matriz de confusión e informe de clasificación.

    TODO para el alumnado:
    - Explica en el informe qué tipo de error consideras más grave.
    - No copies la salida sin comentarla.
    """
    y_pred = model.predict(X_test)
    cm = confusion_matrix(y_test, y_pred)
    report = classification_report(y_test, y_pred)

    print("\nMatriz de confusión:")
    print(cm)
    print("\nClassification report:")
    print(report)


# -----------------------------------------------------------------------------
# 7. PROGRAMA PRINCIPAL
# -----------------------------------------------------------------------------
def main() -> None:
    # -------------------------------------------------------------------------
    # PASO 1. CARGAR Y DIVIDIR LOS DATOS
    # -------------------------------------------------------------------------
    X, y = load_dataset()
    X_train, X_test, y_train, y_test = split_dataset(X, y)

    print("=" * 80)
    print("PRÁCTICA 7 - MODELOS DE ENSAMBLES")
    print("=" * 80)
    print(f"Número de muestras totales: {len(X)}")
    print(f"Número de variables: {X.shape[1]}")
    print(f"Muestras de entrenamiento: {len(X_train)}")
    print(f"Muestras de prueba: {len(X_test)}")

    # -------------------------------------------------------------------------
    # PASO 2. ÁRBOL SIMPLE
    # -------------------------------------------------------------------------
    print("\n" + "-" * 80)
    print("TAREA 1 - ÁRBOL SIMPLE")
    print("-" * 80)

    tree_model = train_decision_tree(X_train, y_train)
    tree_results = evaluate_model(tree_model, X_train, y_train, X_test, y_test)

    print(f"Accuracy en entrenamiento: {tree_results['train_accuracy']:.4f}")
    print(f"Accuracy en test:          {tree_results['test_accuracy']:.4f}")

    print_confusion_and_report(tree_model, X_test, y_test)

    # TODO para el alumnado:
    # Responde en el informe:
    # 1. ¿Hay señales de sobreajuste?
    # 2. ¿Por qué un árbol muy profundo puede memorizar el entrenamiento?

    # -------------------------------------------------------------------------
    # PASO 3. RANDOM FOREST
    # -------------------------------------------------------------------------
    print("\n" + "-" * 80)
    print("TAREA 2 - RANDOM FOREST")
    print("-" * 80)

    n_trees_list = [1, 10, 50, 100]
    rf_results = random_forest_experiment(
        X_train,
        y_train,
        X_test,
        y_test,
        n_trees_list=n_trees_list,
    )

    print("\nResultados Random Forest:")
    print(rf_results)

    plot_random_forest_results(rf_results)

    # Entrenamos un modelo final con 100 árboles para analizarlo mejor.
    rf_model = RandomForestClassifier(n_estimators=100, random_state=RANDOM_STATE)
    rf_model.fit(X_train, y_train)

    print("\nEvaluación Random Forest (100 árboles):")
    print_confusion_and_report(rf_model, X_test, y_test)

    print("\nTop 3 variables más importantes en Random Forest:")
    print(show_top_features(rf_model, list(X.columns), top_k=3))

    # TODO para el alumnado:
    # Responde en el informe:
    # 1. ¿Mejora Random Forest frente al árbol simple?
    # 2. ¿Se estabiliza el resultado al aumentar el número de árboles?
    # 3. ¿Por qué el bagging reduce la varianza del modelo?

    # -------------------------------------------------------------------------
    # PASO 4. BOOSTING
    # -------------------------------------------------------------------------
    print("\n" + "-" * 80)
    print("TAREA 3 - BOOSTING")
    print("-" * 80)

    gb_model, gb_train_time = train_gradient_boosting(X_train, y_train)
    gb_results = evaluate_model(gb_model, X_train, y_train, X_test, y_test)

    print(f"Tiempo de entrenamiento:   {gb_train_time:.4f} s")
    print(f"Accuracy en entrenamiento: {gb_results['train_accuracy']:.4f}")
    print(f"Accuracy en test:          {gb_results['test_accuracy']:.4f}")

    print_confusion_and_report(gb_model, X_test, y_test)

    print("\nTop 3 variables más importantes en Gradient Boosting:")
    print(show_top_features(gb_model, list(X.columns), top_k=3))

    # TODO para el alumnado:
    # Responde en el informe:
    # 1. ¿Qué diferencias observas entre Random Forest y Boosting?
    # 2. ¿Cuál parece entrenar más rápido en tu ejecución?
    # 3. ¿Cuál elegirías si priorizas rendimiento? ¿Y si priorizas simplicidad?

    # -------------------------------------------------------------------------
    # PASO 5. TABLA COMPARATIVA FINAL
    # -------------------------------------------------------------------------
    print("\n" + "-" * 80)
    print("TABLA COMPARATIVA FINAL")
    print("-" * 80)

    comparison_df = pd.DataFrame(
        [
            {
                "Modelo": "Árbol simple",
                "Accuracy test": tree_results["test_accuracy"],
                "Ventajas": "TODO",
                "Desventajas": "TODO",
            },
            {
                "Modelo": "Random Forest",
                "Accuracy test": rf_model.score(X_test, y_test),
                "Ventajas": "TODO",
                "Desventajas": "TODO",
            },
            {
                "Modelo": "Gradient Boosting",
                "Accuracy test": gb_results["test_accuracy"],
                "Ventajas": "TODO",
                "Desventajas": "TODO",
            },
        ]
    )

    print(comparison_df)

    # TODO para el alumnado:
    # Completa esta reflexión final en el informe:
    # "Si tuvieras que desplegar un modelo en una aplicación móvil con poca
    # memoria y CPU, ¿cuál elegirías y por qué?"


if __name__ == "__main__":
    main()

File: P7_Ensambles/practica7_ensambles_alumnos.jl
"""
PRÁCTICA 7: La Inteligencia Colectiva (Modelos de Ensambles)
Asignatura: Introducción al Aprendizaje Automático (IAA) 26/27

Versión en Julia del código base.

Notas importantes
-----------------
- Esta versión utiliza un CSV local con el dataset Breast Cancer Wisconsin.
- El fichero `breast_cancer_wisconsin.csv` debe estar en la misma carpeta que este script.
- El modelo de boosting usado aquí es AdaBoost, que encaja con la práctica
  porque en el guión se permite usar Gradient Boosting o AdaBoost.

Paquetes necesarios
-------------------
using Pkg
Pkg.add(["CSV", "DataFrames", "Plots", "DecisionTree"])

Objetivo
--------
El script ofrece una estructura guiada para que el alumnado implemente,
complete y analice la práctica. No es una solución cerrada: mantiene
comentarios y preguntas orientadas al informe final.
"""

using CSV
using DataFrames
using DecisionTree
using Plots
using Printf
using Random
using Statistics

const RANDOM_STATE = 42
const TEST_SIZE = 0.25
const DATA_PATH = joinpath(@__DIR__, "breast_cancer_wisconsin.csv")
const RF_PLOT_PATH = joinpath(@__DIR__, "random_forest_accuracy_julia.png")


# -----------------------------------------------------------------------------
# 1. CARGA Y PREPARACIÓN DE LOS DATOS
# -----------------------------------------------------------------------------
function load_dataset(path::String=DATA_PATH)
    """
    Carga el dataset desde un CSV local.

    Returns
    -------
    X : Matrix{Float64}
        Variables predictoras.
    y : Vector{String}
        Etiquetas de clase.
    feature_names : Vector{String}
        Nombres de las variables.
    """
    df = CSV.read(path, DataFrame)
    feature_names = String.(names(df)[1:end-1])
    X = Matrix{Float64}(df[:, feature_names])
    y = String.(df[:, :target])
    return X, y, feature_names
end


function stratified_split(
    X::Matrix{Float64},
    y::Vector{String};
    test_size::Float64=TEST_SIZE,
    random_state::Int=RANDOM_STATE,
)
    """
    Divide el dataset en entrenamiento y prueba preservando la proporción de clases.

    TODO para el alumnado:
    - Revisa por qué conviene mantener la proporción de clases.
    - Comprueba cuántas muestras de cada clase quedan en train y en test.
    """
    rng = MersenneTwister(random_state)
    train_idx = Int[]
    test_idx = Int[]

    for cls in unique(y)
        idx = findall(==(cls), y)
        shuffle!(rng, idx)

        n_test = round(Int, length(idx) * test_size)
        n_test = clamp(n_test, 1, length(idx) - 1)

        append!(test_idx, idx[1:n_test])
        append!(train_idx, idx[n_test+1:end])
    end

    shuffle!(rng, train_idx)
    shuffle!(rng, test_idx)

    return X[train_idx, :], X[test_idx, :], y[train_idx], y[test_idx]
end


# -----------------------------------------------------------------------------
# 2. MÉTRICAS Y UTILIDADES DE EVALUACIÓN
# -----------------------------------------------------------------------------
function accuracy_score(y_true::Vector{String}, y_pred::Vector{String})
    return mean(y_true .== y_pred)
end


function confusion_matrix_manual(y_true::Vector{String}, y_pred::Vector{String})
    labels = sort(unique(vcat(y_true, y_pred)))
    label_to_idx = Dict(label => i for (i, label) in enumerate(labels))
    cm = zeros(Int, length(labels), length(labels))

    for (yt, yp) in zip(y_true, y_pred)
        cm[label_to_idx[yt], label_to_idx[yp]] += 1
    end

    return labels, cm
end


function classification_report_df(y_true::Vector{String}, y_pred::Vector{String})
    labels = sort(unique(vcat(y_true, y_pred)))
    rows = NamedTuple[]

    for label in labels
        tp = sum((y_true .== label) .& (y_pred .== label))
        fp = sum((y_true .!= label) .& (y_pred .== label))
        fn = sum((y_true .== label) .& (y_pred .!= label))
        support = sum(y_true .== label)

        precision = (tp + fp) == 0 ? 0.0 : tp / (tp + fp)
        recall = (tp + fn) == 0 ? 0.0 : tp / (tp + fn)
        f1 = (precision + recall) == 0 ? 0.0 : 2 * precision * recall / (precision + recall)

        push!(rows, (
            class=label,
            precision=precision,
            recall=recall,
            f1_score=f1,
            support=support,
        ))
    end

    return DataFrame(rows)
end


function evaluate_model(model, X_train, y_train, X_test, y_test; predictor=predict)
    """
    Evalúa un modelo en entrenamiento y en prueba.

    Returns
    -------
    results : NamedTuple
        Accuracy en entrenamiento y prueba.

    TODO para el alumnado:
    - Añade otras métricas si lo consideras útil.
    - Interpreta la diferencia entre train_accuracy y test_accuracy.
    """
    y_train_pred = String.(predictor(model, X_train))
    y_test_pred = String.(predictor(model, X_test))

    return (
        train_accuracy = accuracy_score(y_train, y_train_pred),
        test_accuracy = accuracy_score(y_test, y_test_pred),
        y_train_pred = y_train_pred,
        y_test_pred = y_test_pred,
    )
end


function print_confusion_and_report(model, X_test, y_test; predictor=predict)
    """
    Imprime matriz de confusión e informe de clasificación simplificado.

    TODO para el alumnado:
    - Explica en el informe qué tipo de error consideras más grave.
    - No copies la salida sin comentarla.
    """
    y_pred = String.(predictor(model, X_test))
    labels, cm = confusion_matrix_manual(y_test, y_pred)
    report = classification_report_df(y_test, y_pred)

    println("\nEtiquetas (filas = real, columnas = predicción): ", labels)
    println("Matriz de confusión:")
    println(cm)

    println("\nClassification report:")
    show(report, allrows=true, allcols=true)
    println()
end


# -----------------------------------------------------------------------------
# 3. TAREA 1: ÁRBOL SIMPLE (BASELINE)
# -----------------------------------------------------------------------------
function train_decision_tree(X_train, y_train)
    """
    Entrena un árbol de decisión sin limitar la profundidad.

    TODO para el alumnado:
    - Comprueba qué ocurre cuando no fijamos max_depth.
    - Relaciona el resultado con el concepto de sobreajuste.
    """
    model = DecisionTreeClassifier()
    DecisionTree.fit!(model, X_train, y_train)
    return model
end


# -----------------------------------------------------------------------------
# 4. TAREA 2: RANDOM FOREST
# -----------------------------------------------------------------------------
function train_random_forest(X_train, y_train; n_trees::Int=100)
    model = RandomForestClassifier(n_trees=n_trees)
    DecisionTree.fit!(model, X_train, y_train)
    return model
end


function random_forest_experiment(X_train, y_train, X_test, y_test, n_trees_list)
    """
    Ejecuta varios experimentos variando el número de árboles.

    TODO para el alumnado:
    - Observa si llega un punto de saturación.
    - Compara estos resultados con el árbol simple.
    """
    rows = NamedTuple[]

    for n_trees in n_trees_list
        start_ns = time_ns()
        model = train_random_forest(X_train, y_train; n_trees=n_trees)
        elapsed = (time_ns() - start_ns) / 1e9

        y_test_pred = String.(DecisionTree.predict(model, X_test))
        test_accuracy = accuracy_score(y_test, y_test_pred)

        push!(rows, (
            n_estimators=n_trees,
            test_accuracy=test_accuracy,
            train_time_seconds=elapsed,
        ))
    end

    return DataFrame(rows)
end


function plot_random_forest_results(results_df::DataFrame; save_path::String=RF_PLOT_PATH)
    """
    Representa la evolución del accuracy en test frente al número de árboles.

    TODO para el alumnado:
    - Cambia el gráfico para mostrar error en lugar de accuracy si lo prefieres.
    - Interpreta la forma de la curva en el informe.
    """
    plt = plot(
        results_df.n_estimators,
        results_df.test_accuracy,
        marker=:circle,
        linestyle=:dash,
        xlabel="n_estimators",
        ylabel="Accuracy en test",
        title="Random Forest: Accuracy en test vs número de árboles",
        legend=false,
        grid=true,
        size=(800, 500),
    )

    savefig(plt, save_path)
    display(plt)
    println("\nGráfica guardada en: $(save_path)")
end


# -----------------------------------------------------------------------------
# 5. TAREA 3: BOOSTING (ADABOOST)
# -----------------------------------------------------------------------------
function train_adaboost(X_train, y_train; n_iterations::Int=100)
    """
    Entrena un modelo AdaBoost de stumps y mide el tiempo de entrenamiento.

    Nota didáctica
    --------------
    En la versión Python del guión se proponía GradientBoostingClassifier o,
    alternativamente, AdaBoostClassifier. Aquí usamos AdaBoostStumpClassifier,
    que es una opción natural dentro de DecisionTree.jl.

    TODO para el alumnado:
    - Interpreta la diferencia entre entrenamiento secuencial (boosting)
      y entrenamiento más independiente de bagging.
    - Compara el comportamiento frente a Random Forest.
    """
    start_ns = time_ns()
    model = AdaBoostStumpClassifier(n_iterations=n_iterations)
    DecisionTree.fit!(model, X_train, y_train)
    elapsed = (time_ns() - start_ns) / 1e9
    return model, elapsed
end


# -----------------------------------------------------------------------------
# 6. IMPORTANCIA DE VARIABLES
# -----------------------------------------------------------------------------
function _importance_vector(model, n_features::Int)
    raw_importance = try
        DecisionTree.impurity_importance(model)
    catch
        DecisionTree.split_importance(model)
    end

    if raw_importance isa AbstractVector
        return Float64.(raw_importance)
    elseif raw_importance isa AbstractDict
        values = zeros(Float64, n_features)
        for (k, v) in raw_importance
            values[Int(k)] = Float64(v)
        end
        return values
    else
        error("No se pudo interpretar el formato de importancia de variables.")
    end
end


function show_top_features(model, feature_names::Vector{String}; top_k::Int=3)
    """
    Devuelve las variables más importantes según el modelo.

    TODO para el alumnado:
    - Comenta si estas variables tienen sentido en el problema.
    - Compara la importancia obtenida en Random Forest y en Boosting.
    """
    importances = _importance_vector(model, length(feature_names))
    df = DataFrame(feature=feature_names, importance=importances)
    sort!(df, :importance, rev=true)
    return first(df, min(top_k, nrow(df)))
end


# -----------------------------------------------------------------------------
# 7. PROGRAMA PRINCIPAL
# -----------------------------------------------------------------------------
function main()
    # -------------------------------------------------------------------------
    # PASO 1. CARGAR Y DIVIDIR LOS DATOS
    # -------------------------------------------------------------------------
    X, y, feature_names = load_dataset()
    X_train, X_test, y_train, y_test = stratified_split(X, y)

    println(repeat("=", 80))
    println("PRÁCTICA 7 - MODELOS DE ENSAMBLES (JULIA)")
    println(repeat("=", 80))
    println("Número de muestras totales: $(size(X, 1))")
    println("Número de variables: $(size(X, 2))")
    println("Muestras de entrenamiento: $(size(X_train, 1))")
    println("Muestras de prueba: $(size(X_test, 1))")

    # -------------------------------------------------------------------------
    # PASO 2. ÁRBOL SIMPLE
    # -------------------------------------------------------------------------
    println("\n" * repeat("-", 80))
    println("TAREA 1 - ÁRBOL SIMPLE")
    println(repeat("-", 80))

    tree_start_ns = time_ns()
    tree_model = train_decision_tree(X_train, y_train)
    tree_train_time = (time_ns() - tree_start_ns) / 1e9
    tree_results = evaluate_model(tree_model, X_train, y_train, X_test, y_test; predictor=DecisionTree.predict)

    @printf("Tiempo de entrenamiento:   %.4f s\n", tree_train_time)
    @printf("Accuracy en entrenamiento: %.4f\n", tree_results.train_accuracy)
    @printf("Accuracy en test:          %.4f\n", tree_results.test_accuracy)

    print_confusion_and_report(tree_model, X_test, y_test; predictor=DecisionTree.predict)

    # TODO para el alumnado:
    # 1. ¿Hay señales de sobreajuste?
    # 2. ¿Por qué un árbol muy profundo puede memorizar el entrenamiento?

    # -------------------------------------------------------------------------
    # PASO 3. RANDOM FOREST
    # -------------------------------------------------------------------------
    println("\n" * repeat("-", 80))
    println("TAREA 2 - RANDOM FOREST")
    println(repeat("-", 80))

    n_trees_list = [1, 10, 50, 100]
    rf_curve_results = random_forest_experiment(X_train, y_train, X_test, y_test, n_trees_list)

    println("\nResultados Random Forest:")
    show(rf_curve_results, allrows=true, allcols=true)
    println()

    plot_random_forest_results(rf_curve_results)

    rf_start_ns = time_ns()
    rf_model = train_random_forest(X_train, y_train; n_trees=100)
    rf_train_time = (time_ns() - rf_start_ns) / 1e9
    rf_results = evaluate_model(rf_model, X_train, y_train, X_test, y_test; predictor=DecisionTree.predict)

    println("\nEvaluación Random Forest (100 árboles):")
    @printf("Tiempo de entrenamiento:   %.4f s\n", rf_train_time)
    @printf("Accuracy en entrenamiento: %.4f\n", rf_results.train_accuracy)
    @printf("Accuracy en test:          %.4f\n", rf_results.test_accuracy)

    print_confusion_and_report(rf_model, X_test, y_test; predictor=DecisionTree.predict)

    println("\nTop 3 variables más importantes en Random Forest:")
    show(show_top_features(rf_model, feature_names; top_k=3), allrows=true, allcols=true)
    println()

    # TODO para el alumnado:
    # 1. ¿Mejora Random Forest frente al árbol simple?
    # 2. ¿Se estabiliza el resultado al aumentar el número de árboles?
    # 3. ¿Por qué el bagging reduce la varianza del modelo?

    # -------------------------------------------------------------------------
    # PASO 4. BOOSTING
    # -------------------------------------------------------------------------
    println("\n" * repeat("-", 80))
    println("TAREA 3 - BOOSTING (ADABOOST)")
    println(repeat("-", 80))

    boost_model, boost_train_time = train_adaboost(X_train, y_train; n_iterations=100)
    boost_results = evaluate_model(boost_model, X_train, y_train, X_test, y_test; predictor=DecisionTree.predict)

    @printf("Tiempo de entrenamiento:   %.4f s\n", boost_train_time)
    @printf("Accuracy en entrenamiento: %.4f\n", boost_results.train_accuracy)
    @printf("Accuracy en test:          %.4f\n", boost_results.test_accuracy)

    print_confusion_and_report(boost_model, X_test, y_test; predictor=DecisionTree.predict)

    println("\nTop 3 variables más importantes en AdaBoost:")
    show(show_top_features(boost_model, feature_names; top_k=3), allrows=true, allcols=true)
    println()

    # TODO para el alumnado:
    # 1. ¿Qué diferencias observas entre Random Forest y Boosting?
    # 2. ¿Cuál parece entrenar más rápido en tu ejecución?
    # 3. ¿Cuál elegirías si priorizas rendimiento? ¿Y si priorizas simplicidad?

    # -------------------------------------------------------------------------
    # PASO 5. TABLA COMPARATIVA FINAL
    # -------------------------------------------------------------------------
    println("\n" * repeat("-", 80))
    println("TABLA COMPARATIVA FINAL")
    println(repeat("-", 80))

    comparison_df = DataFrame(
        Modelo=["Árbol simple", "Random Forest", "AdaBoost"],
        Accuracy_train=[tree_results.train_accuracy, rf_results.train_accuracy, boost_results.train_accuracy],
        Accuracy_test=[tree_results.test_accuracy, rf_results.test_accuracy, boost_results.test_accuracy],
        Tiempo_entrenamiento_s=[tree_train_time, rf_train_time, boost_train_time],
        Ventajas=["TODO", "TODO", "TODO"],
        Desventajas=["TODO", "TODO", "TODO"],
    )

    show(comparison_df, allrows=true, allcols=true)
    println()

    # TODO para el alumnado:
    # Completa esta reflexión final en el informe:
    # "Si tuvieras que desplegar un modelo en una aplicación móvil con poca
    # memoria y CPU, ¿cuál elegirías y por qué?"
end


main()
